### ISSUE TITLE
Reduce's result type not respecting widening coming from block

### ISSUE THREAD

**Issue by vinistock** [sorbet/sorbet#6153]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0A%5B%5D.reduce%28true%29%20do%20%7Cresult%2C%20item%7C%0A%20%20if%20item.should_check%3F%0A%20%20%20%20result%20%26%26%20item.do_something%0A%20%20else%0A%20%20%20%20false%0A%20%20end%0Aend)

```ruby
# typed: true

[].reduce(true) do |result, item|
  if item.should_check?
    result && item.do_something
  else
    false
  end
end
```

#### Observed output

```
editor.rb:5: This code is unreachable https://srb.help/7006
     5 |    result && item.do_something
            ^^^^^^
    editor.rb:5: This condition was always truthy (TrueClass)
     5 |    result && item.do_something
            ^^^^^^
  Got TrueClass originating from:
    editor.rb:3:
     3 |[].reduce(true) do |result, item|
                            ^^^^^^
Errors: 1
```

#### Expected behavior

Likely related to #6087

When using `reduce`, we are setting the type of the first block argument (result in the example) to be the same type as the one received in the positional argument of `reduce`.

In this case, `true` is a `TrueClass` and therefore `result` is also a `TrueClass`. However, that's not correct given that `result` can be mutated by the block's return. In this case, `result` should be a `T::Boolean`, given that the block may return either `true` or `false`.

I think this is mostly a problem when using `reduce` where the type may be widened depending on the block's return. Here's another [example](https://sorbet.run/#%23%20typed%3A%20true%0A%0A%5B1.0%2C%202.0%2C%203.0%5D.reduce%282%29%20do%20%7Cresult%2C%20item%7C%0A%20%20%23%20It's%20Integer%20because%20of%20the%20initial%20value%20of%202%2C%20but%20after%20the%20first%20iteration%0A%20%20%23%20it%20becomes%20a%20Float%0A%20%20T.reveal_type%28result%29%0A%20%20result%20*%20item%0Aend%0A) using numbers.

**Comment by jez** [comment]
This is largely the same issue of Sorbet requiring explicit type annotations when changing the type of a variable in a loop.

```ruby
x = false
if T.unsafe(true)
  x = true
end
```

e.g., this will be a pinning error, and needs to be fixed with a `T.let`:

```ruby
x = T.let(true, T::Boolean)
if T.unsafe(true)
  x = false
end
```


Similarly, the `reduce` needs a `T.let` annotation to widen the type:

```ruby
[].reduce(T.let(true, T::Boolean)) do |result, item|
  if item.should_check?
    result && item.do_something
  else
    false
  end
end
```

This limitation seems acceptable not only by analogy with the pinning error, but also because most cases where this comes up have an easier work around, for example:

```ruby
[].all? do |item|
  if item.should_check?
    item.do_something
  else
    false
  end
end
```

And the float example:

```ruby
[1.0, 2.0, 3.0].reduce(2.0) do |result, item|
  result * item
end
```

Until the day that Sorbet removes the restriction that it type checks every instruction only once, it will not be able to use any information about how the block typechecks to retroactively change a decision about how it has typechecked arguments to the function.

**Comment by vinistock** [comment]
Oh, I hadn't realized we could just `T.let` the initial value of `reduce`. I thought there'd be no way of properly getting the right type on the block.

I think it's fine then, I agree that this is the same ergonomics people are used to. I'll close the issue.

**Comment by CodingAnarchy** [comment]
I just ran into this in the case of attempting a `reduce(false) { |x| x || y }` block, which is the mirror image here. It took a while to find this closed issue indicating I needed to tell Sorbet that as a block argument, the value is a mutatable Boolean, not a `FalseClass` value. The `code is unreachable` error seemed particularly incorrect, as the `falsey` value would mean that the block would fall though, and everything would be executed every time (if it really was always `false || ....`).

If there can be no move to automatically widen the type, particularly of the boolean value in this context, I would suggest more prominent documentation. Based on [the documentation for T::Boolean](https://sorbet.org/docs/class-types#booleans), I somewhat expected that Sorbet automatically inferred `T::Boolean` when the code said `x = true`.
